src/link.cpp

```diff
@@ -1278,11 +1358,27 @@ static void construct_linker_job_coff(LinkJob *lj) {
             continue;
         }
         if (link_lib->provided_explicitly) {
-            if (lj->codegen->zig_target->abi == ZigLLVM_GNU) {
-                Buf *arg = buf_sprintf("-l%s", buf_ptr(link_lib->name));
-                lj->args.append(buf_ptr(arg));
-            }
-            else {
+            if (target_abi_is_gnu(lj->codegen->zig_target->abi)) {
+                static const char *test_names[3] = { "%s\\lib%s.dll.a", "%s\\lib%s.a", nullptr };
+                bool exists = false;
+
+                for (size_t i = 0; test_names[i] != nullptr && !exists; i++) {
+                    for (size_t j = 0; j < g->lib_dirs.length; j += 1) {
+                        const char *lib_dir = g->lib_dirs.at(j);
+                        Buf *test_path = buf_sprintf(test_names[i], lib_dir, buf_ptr(link_lib->name));
+                        if (os_file_exists(test_path, &exists) != ErrorNone) {
```

Comment:
These checks should be during libc detection in `libc_installation.cpp`. Once we get to link.cpp, we should have resolved all the paths we need to link.

Reply:
this is here to make sure that the library that zig is trying to link to actually exists. it's also here because the compiler doesn't have a way (that i know of) to tell you if the library being linked is dynamic or static so this implementation looks for a dynamic library with the given name, then a static one and chooses the first one it finds. this could be fixed by just adding `.dll` to link library name, but i'm not sure. also i think zig telling you that the library doesn't exist is a lot nicer than a cascade of linker errors.

Reply:
In what situation is it possible that some mingw installations have only the .dll files but other mingw installations have only the .a files? Can't we just pick one (.a files would be my preference) and have that always work?

If this logic is needed, it's the first libc among many (msvc, glibc, musl, freebsd, macos) that has this requirement. Is that really the case?

Reply:
i solved this by preferring to link static libraries. if you want to link dynamic library you  have to add the `.dll` suffix to the lib name. so `--library SDL2` for static and `--library SDL2.dll` for dynamic. this should probably be added to some docs but i'm not sure if there is (yet) a place for docs on things like linking.